src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+            self.as_mut_ptr(),
+            len,
+        )
+    }
+
+    /// Initialize the buffer with a copy from another (already initialized)
+    /// buffer.
+    ///
+    /// It returns a read-writable slice to the initialized bytes for
+    /// convenience (automatically
+    /// [`assume_init`][`Out::assume_init`]-ed).
+    ///
+    /// # Panic
+    ///
+    /// The function panics if the slices' lengths are not equal.
+    ///
+    /// # Guarantees
+    ///
+    /// A non-`panic!`king return from this function **guarantees that the input
+    /// slice has been (successfully) initialized**, and that it is thus then
+    /// sound to `.assume_init()`.
+    ///
+    /// It also guarantees that the returned slice does correspond to the input
+    /// slice (_e.g._, for [`crate::ReadIntoUninit`]'s safety guarantees).
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use ::core::mem::{self, MaybeUninit};
+    /// use ::uninit::prelude::*;
+    ///
+    /// let mut array = uninit_array![_; 13];
+    /// assert_eq!(
+    ///     array.as_out().copy_from_slice(b"Hello, World!"),
+    ///     b"Hello, World!",
+    /// );
+    /// // we can thus soundly `assume_init` our array:
+    /// let array = unsafe {
+    ///     mem::transmute::<
+    ///         [MaybeUninit<u8>; 13],
+    ///         [            u8 ; 13],
+    ///     >(array)
+    /// };
+    /// assert_eq!(
+    ///     array,
+    ///     *b"Hello, World!",
+    /// );
+    /// ```
+    pub
+    fn copy_from_slice (
+        mut self: Out<'out, [T]>,
+        source_slice: &'_ [T],
+    ) -> &'out mut [T]
+    where
+        T : Copy,
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - Writing to `self.0` is fine since `source_slice` only
+            //     contains initialized elements.
+            //
+            //   - the `copy_nonoverlapping()` call guarantees that the buffer
+            //     has been initialized.
+            self.r()
+                .as_uninit_mut()
+                .copy_from_slice(
+                    <[MaybeUninit<T>]>::from_ref(source_slice)
+                )
+            ;
+            self.assume_init()
+        }
+    }
+
+    /// Fill the buffer with values from a function mapping indices to values.
+    ///
+    /// # Guarantees
+    ///
+    /// `factory` is guaranteed to be called in the `0 .. self.len()` order.
+    #[inline]
+    pub
+    fn fill_with (
+        mut self: Out<'out, [T]>,
+        mut factory: impl FnMut(usize) -> T,
+    ) -> &'out mut [T]
+    {
+        self.iter_out()
+            .enumerate()
+            .for_each(|(i, at_dst)| { at_dst.write(factory(i)); })
+        ;
+        unsafe {
+            // Safety: The `len` values of the buffer have been initialized
+            self.assume_init()
+        }
+    }
+
+    /// `.reborrow().into_iter()`
+    #[inline]
+    pub
+    fn iter_out<'reborrow> (self: &'reborrow mut Out<'out, [T]>)
```

Comment:
I would add `IntoIterator` for `&'self mut Out<'out>` instead.

Reply:
I think having both makes sense, similar to `<[_]>::iter_mut` and `<&'_ mut [_]>::into_iter`